javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/context/FieldBackedProvider.java

```diff
@@ -881,21 +881,6 @@ public Object get(Object object) {
       return realGet(object);
     }
 
-    @Override
-    public void setIfNull(Object object, Object fieldValue) {
```

Comment:
Now if we could also get rid of `computeIfNull`. There seems to be a small difference between `set` and `setIfNull`. `set` just writes the field, `setIfNull` synchronizes around setting the field. As `get` doesn't synchronize it probably wasn't correct anyway from the concurrency side. Should we make the field volatile to make it at least a bit safer to use? For `computeIfNull` should we synchronize at all, as `set` doesn't synchronize we can't get reasonable multi thread behaviour from this anyway. Or should `set` also synchronize?

Reply:
> Should we make the field volatile to make it at least a bit safer to use?

It wasn't? 😱 

We definitely should! Any changes to those fields should definitely be visible in other threads - they're often used to pass context between different threads.

> Now if we could also get rid of `computeIfNull`.

I guess we could, but it is useful in those several places that use it - it makes the code a bit clearer.

> For `computeIfNull` should we synchronize at all, as `set` doesn't synchronize we can't get reasonable multi thread behaviour from this anyway.

I don't think we care about atomicity in the case of `VirtualField` (it has even proved to be a problem in certain situations: #4188), the visibility of changes is what's really important. I think I'd be fine with dropping the synchronization and just making the field volatile.

Reply:
I created 2 tracking issues to continue these discussions 👍 #4268 and #4269